File: p4_routing/p/send.go
package main

import (
	"log"
	"os"
	"strings"

	"github.com/streadway/amqp"
)

func failOnError(err error, msg string) {
	if err != nil {
		log.Fatalf("%s, %s", msg, err)
	}
}

func bodyFrom(args []string) string {
	var s string
	if (len(args) < 3) || os.Args[2] == "" {
		s = "hello"
	} else {
		s = strings.Join(args[2:], " ")
	}

	return s
}

func severityFrom(args []string) string {
	var s string
	// log.Printf("args[0]: %s, args[1]:%s, args[2]:%s", args[0], args[1], args[2])
	log.Println(args)
	if (len(args) < 2) || (args[1] == "") {
		s = "info"
	} else {
		s = args[1]
	}

	return s
}

func main() {
	conn, err := amqp.Dial("amqp://guest:guest@localhost:5672/")
	failOnError(err, "Failed to connect to RabbitMQ")
	defer conn.Close()

	ch, err := conn.Channel()
	failOnError(err, "Failed to open a channel")
	defer ch.Close()

	err = ch.ExchangeDeclare(
		"direct_logs",
		"direct",
		true,
		false,
		false,
		false,
		nil,
	)
	failOnError(err, "Failed to declare an exchange")

	body := bodyFrom(os.Args)
	err = ch.Publish(
		"direct_logs",
		severityFrom(os.Args),
		false,
		false,
		amqp.Publishing{
			ContentType: "text/plain",
			Body:        []byte(body),
		},
	)
	failOnError(err, "Failed to publish a message")

	log.Printf("[x] Sent %s", body)
}

File: p4_routing/p/send.py
#!/usr/bin/env python
# _*_coding:utf-8_*_

"""
@Datetime :   4/16/20 10:49 PM
@Author   :   Fangyang
"""
import sys

import pika

connection = pika.BlockingConnection(pika.ConnectionParameters("localhost"))
channel = connection.channel()

channel.exchange_declare(
    exchange='direct_logs',
    exchange_type='direct',
    durable=True,  # golang 那边设置了这个参数, 如果不设置为True, 生成的exchange是同名, 但是是不一致的, py和go先后执行会报错
)

# 让rabbitmq自动生成queue
# channel.queue_declare(
#     queue="task_queue",
#     durable=True  # 队列持久化, 在RabbitMQ启动时还在
# )

severity = sys.argv[1] if len(sys.argv) > 1 else 'info'
msg = ' '.join(sys.argv[2:]) or "hello......"
channel.basic_publish(
    exchange="direct_logs",
    routing_key=severity,
    body=msg,
    # properties=pika.BasicProperties(
    #     delivery_mode=2  # message persistent
    # )
)
print(f"[*] Sent {msg}")
connection.close()


if __name__ == "__main__":
    pass
